### PR TITLE
add keyboard [syn_c] - BDN9+Proton-C with working arm pwm support

### PR DESCRIPTION
**Pull request by kratsyn**
This is basically a keeb.io BDN9 with a Proton-C strapped to it.

## Description
- native arm build (!CTPC)
- full pwm support on led output (B0)
- partially generalized arm backlight led implementation

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/syn_c/syn_c.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+typedef union {
+	uint16_t raw;
+	struct {
+		bool 	enable			:1;
+		uint8_t level			:5;
+		bool 	breath_enable	:1;
+	};
+} kb_cfg_backlight;
```
Comment (fauxpark) [inline keyboards/syn_c/syn_c.h]: You shouldn't need to store your own backlight config in the keyboard section of EEPROM - the core backlight section recently received a breathing flag and is now basically identical to this struct.

Comment (kratsyn) [reply keyboards/syn_c/syn_c.h]: I'll make changes to meet the structure standard and refactor later once the arm backlight split has occurred.

Comment (kratsyn) [reply keyboards/syn_c/syn_c.h]: scratch that, done.

File: keyboards/keebio/bdn9/proton_c/eeconfig.h [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+uint8_t eeconfig_read_debug(void);
+void eeconfig_update_debug(uint8_t val);
+
+uint8_t eeconfig_read_default_layer(void);
+void eeconfig_update_default_layer(uint8_t val);
+
+uint8_t eeconfig_read_keymap(void);
+void eeconfig_update_keymap(uint8_t val);
+
+uint8_t eeconfig_read_backlight(void);
```
Comment (fauxpark) [inline keyboards/keebio/bdn9/proton_c/eeconfig.h]: Good point about the ifdef locking these out - theoretically these shouldn't get compiled in anyway if the backlight feature is disabled... and in fact the ifdef is not present in the .c file.

Comment (fauxpark) [reply keyboards/keebio/bdn9/proton_c/eeconfig.h]: I think if you turn on `BACKLIGHT_ENABLE` and add `BACKLIGHT_CUSTOM_DRIVER = yes` in rules.mk, this file is not needed.

Comment (kratsyn) [reply keyboards/keebio/bdn9/proton_c/eeconfig.h]: oh nice. thanks for the tip.

Comment (kratsyn) [reply keyboards/keebio/bdn9/proton_c/eeconfig.h]: the problem there is that it includes tmk's backlight.h which has collisions with this implementation.

Comment (fauxpark) [reply keyboards/keebio/bdn9/proton_c/eeconfig.h]: You shouldn't need to reimplement the functions in there, only the actual implementation from quantum.c - backlight.h just manages the EEPROM and backlight struct, really.

File: keyboards/keebio/bdn9/proton_c/config.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+#define LOCKING_SUPPORT_ENABLE
+#define LOCKING_RESYNC_ENABLE
+
+#define ENCODERS_PAD_A { B6, A0 }
+#define ENCODERS_PAD_B { B7, A1 }
+#define ENCODER_RESOLUTION 4
+
+#define TAP_CODE_DELAY 20 
+#define TAPPING_TERM 300
```
Comment (zvecr) [inline keyboards/keebio/bdn9/proton_c/config.h]: Defaults should be used, so users get the chance to override.
```suggestion
```

File: keyboards/keebio/bdn9/proton_c/config.h [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+#define LOCKING_SUPPORT_ENABLE
+#define LOCKING_RESYNC_ENABLE
+
+#define ENCODERS_PAD_A { B6, A0 }
+#define ENCODERS_PAD_B { B7, A1 }
+#define ENCODER_RESOLUTION 4
+
+#define TAP_CODE_DELAY 20
```
Comment (zvecr) [inline keyboards/keebio/bdn9/proton_c/config.h]: Defaults should be used, so users get the chance to override.
```suggestion
```

**Comment by kratsyn** [comment]
It looks like the voeschu-browser and voeschu-ksp keymaps will be broken until i can shoehorn @awkannan's SPI implementation for ws2812 into the build. There is an added complication there as the bdn9 pin on the proton_c (A9) does not support SPI. i'll have to patch it to a more suitable pad and find a led strip for testing.

On that note, i do have a nagging question related to this exercise...

Would it not be reasonable to allow us the ability to bypass quantum methods in favor of local implementations by supporting *enable=no/*custom_driver=yes combinations? 

It would allow keyboard maintainers the option to override quantum methods while still allowing supporting functionality from eeconfig and similar features.

**Comment by zvecr** [comment]
@kratsyn https://github.com/qmk/qmk_firmware/pull/6301 is currently targeting f303 board like the Proton C, not the SPI variant.

**Comment by kratsyn** [comment]
it is not specific to this pr, just an observation and the reason i'm defining eeprom.h locally. Perhaps i'm missing something obvious.

**Comment by fauxpark** [comment]
`BACKLIGHT_CUSTOM_DRIVER` is exactly what you want - it stops the backlight functions in quantum.c being defined, which is where the actual PWM control happens, but you are overriding the functions in tmk_core/backlight.c that aren't meant to be overridden.

**Comment by kratsyn** [comment]
feel a bit dense having not seen that sooner...

File: keyboards/keebio/bdn9/keymaps/kratsyn-dbg/config.h [hunk 5]
```diff
@@ -0,0 +1,2 @@
+#define TAP_CODE_DELAY	20
```
Comment (zvecr) [inline keyboards/keebio/bdn9/keymaps/kratsyn-dbg/config.h]: Headers should have #pragma once guards
```suggestion
#pragma once
#define TAP_CODE_DELAY	20
```

File: keyboards/keebio/bdn9/keymaps/kratsyn-dbgled/config.h [hunk 6]
```diff
@@ -0,0 +1,2 @@
+#define TAP_CODE_DELAY	20
```
Comment (zvecr) [inline keyboards/keebio/bdn9/keymaps/kratsyn-dbgled/config.h]: Headers should have #pragma once guards
```suggestion
#pragma once
#define TAP_CODE_DELAY	20
```

File: keyboards/keebio/bdn9/keymaps/kratsyn/config.h [hunk 7]
```diff
@@ -0,0 +1,2 @@
+#define TAP_CODE_DELAY	20
```
Comment (zvecr) [inline keyboards/keebio/bdn9/keymaps/kratsyn/config.h]: Headers should have #pragma once guards
```suggestion
#pragma once
#define TAP_CODE_DELAY	20
```

File: keyboards/keebio/bdn9/proton_c/proton_c.h [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+	KB1, KB2, KB3, \
+	KC1, KC2, KC3 \
+) \
+{ \
+	{ KA1, KA2, KA3 }, \
+	{ KB1, KB2, KB3 }, \
+	{ KC1, KC2, KC3 } \
+}
+
+typedef union {
```
Comment (zvecr) [inline keyboards/keebio/bdn9/proton_c/proton_c.h]: As far as I can tell, all the "status" handling doesnt amount to anything and should be removed.

File: keyboards/keebio/bdn9/proton_c/config.h [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+#define	BACKLIGHT_ARM_PIN		B0			// temporary until PR#6487 is included with support for custom drivers
+#define BACKLIGHT_PWM_DRIVER	&PWMD3		// PWM Driver Channel (TIMX), must be enabled in mcuconf.h)
+#define BACKLIGHT_PWM_CHANEL	2			// PWM Timing Channel (TIMX_CHN, where N is the timing channel)
+
+#define BACKLIGHT_LEVELS 5 
+#define BACKLIGHT_BREATHING
+#define BACKLIGHT_BREATHING_STEPS 128
+#define BREATHING_PERIOD 6 
+
+//// PR#6301
```
Comment (zvecr) [inline keyboards/keebio/bdn9/proton_c/config.h]: Given #6301 has not defined the configuration options yet, this is really putting the cart before the horse.

**Comment by zvecr** [comment]
Again, I would reiterate, we have #6487 for backlight in the works, investing time in a bespoke backlight solution for this PR seems counter productive.

**Comment by kratsyn** [comment]
simple fixes that are implemented while doing other things is not time wasted imo. This will now work regardless of the time to resolve 6507. Once that is in place, it can tested upon the assignment of a couple of config.h defines.

**Comment by yanfali** [comment]
@kratsyn awesome work, it's always amazing when the community creates
awesome stuff using QMK as a base.  We have some concerns that
prevent us from accepting this PR as is.

We had a similar situation when split support was merged many moons
ago.  It became the reference implementation and from that were
born a thousand forks. We've finally gotten some of that tamed now
with split-common, but it took a long while and ended up with a lot
of unnecessary duplication and extra work for the maintainers.

@zvecr is working on this already and we would love to have you
contribute your talent and energy to this in-progress work,
addressing your needs and concerns directly, rather than have a potentially
confusing similar standalone feature?

**Comment by kratsyn** [comment]
That was my intention after gauging what that might first be like. I've simplified the underlying implementation greatly through use while @zvecr has done a good job of splitting out the logic to support a more appropriate driver location for any implementation. I would not be opposed to merging efforts.

**Pull request closed without merging**